lib/units/provider/index.ts

```diff
@@ -0,0 +1,420 @@
+    const lists = {
+        all: new Set(),
+        ready: new Set(),
+        waiting: new Set()
+    }
+    let totalsTimer: NodeJS.Timeout
+
+    // To make sure that we always bind the same type of service to the same
+    // port, we must ensure that we allocate ports in fixed groups.
+    let ports = options.ports.slice(0, options.ports.length - options.ports.length % 4)
+
+    // Information about total devices
+    const totals = () => {
+        if (lists.waiting.size) {
+            log.info('Providing %d of %d device(s); waiting for "%s"', lists.ready.size, lists.all.size, Array.from(lists.waiting).join('", "'))
+            delayedTotals()
+        }
+        else if (lists.ready.size < lists.all.size) {
+            log.info('Providing all %d of %d device(s); ignoring not ready: "%s"', lists.ready.size, lists.all.size, _.difference(Array.from(lists.all), Array.from(lists.ready)).join('", "'))
+        }
+        else {
+            log.info('Providing all %d device(s)', lists.all.size)
+        }
+    }
+
+    const delayedTotals = () => {
+        clearTimeout(totalsTimer)
+        totalsTimer = setTimeout(totals, 10000)
+    }
+
+    // Output
+    const push = zmqutil.socket('push')
+    try {
+        await Promise.all(options.endpoints.push.map(async(endpoint) => {
+            const records = await srv.resolve(endpoint)
+            return await srv.attempt(records, (record) => {
+                log.info('Sending output to "%s"', record.url)
+                push.connect(record.url)
+                return Promise.resolve(true)
+            })
+        }))
+    }
+    catch (err) {
+        log.fatal('Unable to connect to push endpoint', err)
+        lifecycle.fatal()
+    }
+
+    // Input
+    const sub = zmqutil.socket('sub')
+    try {
+        await Promise.all(options.endpoints.sub.map(async(endpoint) => {
+            const records = await srv.resolve(endpoint)
+            return await srv.attempt(records, (record) => {
+                log.info('Receiving input to "%s"', record.url)
+                sub.connect(record.url)
+                return Promise.resolve(true)
+            })
+        }))
+    }
+    catch (err) {
+        log.fatal('Unable to connect to sub endpoint', err)
+        lifecycle.fatal()
+    }
+
+
+    ;[solo].forEach(function(channel) {
+        log.info('Subscribing to permanent channel "%s"', channel)
+        sub.subscribe(channel)
+    })
+
+    // This can happen when ADB doesn't have a good connection to
+    // the device
+    const isWeirdUnusableDevice = (device: Device) =>
+        device.id === '????????????'
+
+    // Check whether the device is remote (i.e. if we're connecting to
+    // an IP address (or hostname) and port pair).
+    const isRemoteDevice = (device: Device) =>
+        device.id.includes(':')
+
+    // Helper for ignoring unwanted devices
+    const filterDevice = (listener: (device: Device) => any | Promise<any>) =>
```

Comment:
also we need to implement check for device with same serial as already exists in provider
and if its physical - we need to implement custom logic in future